arbnode/transaction_streamer.go

```diff
@@ -218,68 +195,100 @@ func deleteStartingAt(db ethdb.Database, batch ethdb.Batch, prefix []byte, minKe
+func (s *TransactionStreamer) reorg(batch ethdb.Batch, count arbutil.MessageIndex, newMessages []arbostypes.MessageWithMetadata) error {
 	if count == 0 {
-		return nil, errors.New("cannot reorg out init message")
+		return errors.New("cannot reorg out init message")
+	}
+	lastDelayedSeqNum, err := s.getPrevPrevDelayedRead(count)
+	if err != nil {
+		return err
+	}
+	var oldMessages []*arbostypes.MessageWithMetadata
+
+	targetMsgCount, err := s.GetMessageCount()
+	if err != nil {
+		return err
+	}
+	config := s.config()
+	maxResequenceMsgCount := count + arbutil.MessageIndex(config.MaxReorgResequenceDepth)
+	if config.MaxReorgResequenceDepth >= 0 && maxResequenceMsgCount < targetMsgCount {
+		log.Error(
+			"unable to re-sequence all old messages because there are too many",
+			"reorgingToCount", count,
+			"removingMessages", targetMsgCount-count,
+			"maxReorgResequenceDepth", config.MaxReorgResequenceDepth,
+		)
+		targetMsgCount = maxResequenceMsgCount
 	}
-	var oldMessages []*arbstate.MessageWithMetadata
-	if s.reorgSequencing {
-		targetMsgCount, err := s.GetMessageCount()
+	for i := count; i < targetMsgCount; i++ {
+		oldMessage, err := s.GetMessage(i)
 		if err != nil {
-			return nil, err
+			log.Error("unable to lookup old message for re-sequencing", "position", i, "err", err)
+			break
 		}
-		config := s.config()
-		maxResequenceMsgCount := count + arbutil.MessageIndex(config.MaxReorgResequenceDepth)
-		if config.MaxReorgResequenceDepth >= 0 && maxResequenceMsgCount < targetMsgCount {
-			log.Error(
-				"unable to re-sequence all old messages because there are too many",
-				"reorgingToCount", count,
-				"removingMessages", targetMsgCount-count,
-				"maxReorgResequenceDepth", config.MaxReorgResequenceDepth,
-			)
-			targetMsgCount = maxResequenceMsgCount
+
+		if oldMessage.Message == nil || oldMessage.Message.Header == nil {
+			continue
 		}
-		for i := count; i < targetMsgCount; i++ {
-			oldMessage, err := s.GetMessage(i)
-			if err != nil {
-				log.Error("unable to lookup old message for re-sequencing", "position", i, "err", err)
-				break
+
+		header := oldMessage.Message.Header
+
+		if header.RequestId != nil {
+			// This is a delayed message
+			delayedSeqNum := header.RequestId.Big().Uint64()
+			if delayedSeqNum+1 != oldMessage.DelayedMessagesRead {
+				log.Error("delayed message header RequestId doesn't match database DelayedMessagesRead", "header", oldMessage.Message.Header, "delayedMessagesRead", oldMessage.DelayedMessagesRead)
+				continue
+			}
+			if delayedSeqNum != lastDelayedSeqNum {
+				// This is the wrong position for the delayed message
+				continue
+			}
+			lastDelayedSeqNum++
+
+			if s.inboxReader != nil {
+				// Verify that the delayed message we're re-sequencing matches what we have after the reorg
+				haveDelayedMessage, err := s.inboxReader.tracker.GetDelayedMessage(delayedSeqNum)
```

Comment:
Is the inbox tracker's database committed to at this point? I'm worried that at this point in the function call chain the database batch removing the old delayed messages might not have been committed yet.

Reply:
Perhaps we could have a variant of this method where we pass in the database batch we're in and it queries against that?